CIPs/cip-137.md

```diff
@@ -18,190 +18,239 @@ An API for reading and writing events on Ceramic event streams.
+Content-Type:  `application/vnd.ipld.car`
 
-**Returns:**
+A CAR file where the root is an *Event* as described below.
 
-- *events* <string> - a base64 encoded CAR file containing the event blocks
+```verilog
+// https://developers.ceramic.network/protocol/streams/event-log/#data-event
+type Event InitEvent | DataEvent | TimeEvent
+```
 
----
+**JSON:**
 
+Content-Type:  `application/json`
 
+A possible convenience method for adding an event by only submitting a JWT.
 
-### `ceramic_putRawEvent`
+Only possible once we’ve migrated to use *Varsig* and can create events as plain signed JWTs.
 
-Add an event to a stream.
+```javascript
+{
+  event: "<jwt>"
+}
+```
 
-**Params:**
+#### Returns:
 
-- *event* <string> - a base64 encoded CAR file containing the event
 
-**Returns:**
 
-- *success* <boolean> - true if the event was added correctly
+**JSON:**
 
----
+- *eventid* <string> - the eventId of this event
 
 
 
-### `ceramic_putEvent`
+---
 
-Convenience method for adding an event by only submitting a JWT.
+### GET `/stats/active-ranges`
 
-Only possible once we’ve migrated to use *Varsig* and can create events as plain signed JWTs.
+A status endpoint that returns all ranges that this node is actively monitoring.
 
-**Params:**
+#### Returns:
 
-- *event* <string> - a jwt containing the event
+* *ranges* <array<object>> - a list of all subscriptions
+  * *reconRange* <Array<string>> - a Recon message, *`[eventid or ahash]`*
+  * *offset* <integer> - the total number of eventId bytes observed in this range
 
-**Returns:**
 
-- *success* <boolean> - true if the event was added correctly
 
 
 ## Rationale
 <!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
 
-The choice of using json-rpc enables the API to work over multiple different transports. For example, HTTP, web sockets, or event the iframe postMessage api. The latter could be useful in the future if one want's to run a node inside of an iframe.
+The choice of using OpenAPI enables ease of integration and content types to more easily send and respond using binary formats such as CAR files. It's also a common pattern to have service workers serve http requests opening up for future browser native implementations of the Ceramic protocol.
 
 There are two main ways of polling for new events. Using recon and a more simple poll. This enables advanced developers to get full control with recon, while a simple poll API based on a single event counter is much easier to use for most developers. The disadvantage of the latter is that events will be received in the order that the node received them, not in the global order of the network.
```

Comment:
how is this easier for most developers?  How would a developer actually use this poll API?  How would **we** even use it?  If you get a newer event before an older one, it seems like most of the time you can't use that event at all until you get the earlier one anyway, so either you need to go out of band to fetch it, or wait for it to come in here.  Seems like it'd be better to just have the node do that for you

Reply:
OOOOOHHHHH

Well yes that's a very different thing! That is not at all clear from this document however.  Yes, across streams events can come in any order.  But within a stream we should not return a newer event until we've received an earlier one.

Reply:
(mod events that come in on a different branch of history, which we'll still need to think about how we handle somehow)

Reply:
I think in this case clients would have to do consolidation as they get the events using the *eventHeight* from the EventId.

Fun side note: From a data structure perspective you could actually think of all MIDs that share the same Model as a single stream where the Model InitEvent is a kind of InitEvent for all MIDs.